ui/src/app/shared/use-editable-resource.ts

```diff
@@ -0,0 +1,22 @@
+import {useState} from 'react';
+
+/**
+ * useEditableResource is a hook to manage the state of a resource that be edited and updated.
+ */
+export function useEditableResource<T>(initial?: T): [T, boolean, React.Dispatch<T>, (value: T) => void] {
+    const [value, setValue] = useState<T>(initial);
+    const [initialValue, setInitialValue] = useState<T>(initial);
+
+    // TODO: Fix this so that it handles object comparison properly.
+    // Currently, this returns true if you make a change and immediately undo it, or if you save your changes.
+    // This could be solved using "const edited = JSON.stringify(value) !== JSON.stringify(initialValue)",
+    // but that has a performance penalty.
```

Comment:
I don't think this is a TODO if we've intentionally decided _not_ to do it. We can just say notate we decided not to do it for performance purposes of a latency sensitive input

```suggestion
    // this is pure ref equality instead of deep object comparison etc as that has performance implications and inputs are latency sensitive
```

Reply:
IMO, this is worth fixing, but I went ahead and updated the comment in cf9a3f383d51dc74fc6e8edd0f86b9f413e14d52

Reply:
Can continue that discussion in the [above thread](https://github.com/argoproj/argo-workflows/pull/13593#discussion_r1780324389) then, which is more specifically perf related